contracts/base/ERC7739Validator.sol

```diff
@@ -0,0 +1,233 @@
+    ///   that will call `isValidSignature` on this account.
+    ///
+    /// - `ACCOUNT_DOMAIN_SEPARATOR`: The domain separator of this account.
+    ///   See: `EIP712._domainSeparator()`.
+    /// __________________________________________________________________________________________
+    ///
+    /// For the `TypedDataSign` workflow, the final hash will be:
+    /// ```
+    ///     keccak256(\x19\x01 ‖ APP_DOMAIN_SEPARATOR ‖
+    ///         hashStruct(TypedDataSign({
+    ///             contents: hashStruct(originalStruct),
+    ///             name: keccak256(bytes(eip712Domain().name)),
+    ///             version: keccak256(bytes(eip712Domain().version)),
+    ///             chainId: eip712Domain().chainId,
+    ///             verifyingContract: eip712Domain().verifyingContract,
+    ///             salt: eip712Domain().salt
+    ///             extensions: keccak256(abi.encodePacked(eip712Domain().extensions))
+    ///         }))
+    ///     )
+    /// ```
+    /// where `‖` denotes the concatenation operator for bytes.
+    /// The order of the fields is important: `contents` comes before `name`.
+    ///
+    /// The signature will be `r ‖ s ‖ v ‖
+    ///     APP_DOMAIN_SEPARATOR ‖ contents ‖ contentsType ‖ uint16(contentsType.length)`,
+    /// where `contents` is the bytes32 struct hash of the original struct.
+    ///
+    /// The `APP_DOMAIN_SEPARATOR` and `contents` will be used to verify if `hash` is indeed correct.
+    /// __________________________________________________________________________________________
+    ///
+    /// For the `PersonalSign` workflow, the final hash will be:
+    /// ```
+    ///     keccak256(\x19\x01 ‖ ACCOUNT_DOMAIN_SEPARATOR ‖
+    ///         hashStruct(PersonalSign({
+    ///             prefixed: keccak256(bytes(\x19Ethereum Signed Message:\n ‖
+    ///                 base10(bytes(someString).length) ‖ someString))
+    ///         }))
+    ///     )
+    /// ```
+    /// where `‖` denotes the concatenation operator for bytes.
+    ///
+    /// The `PersonalSign` type hash will be `keccak256("PersonalSign(bytes prefixed)")`.
+    /// The signature will be `r ‖ s ‖ v`.
+    /// __________________________________________________________________________________________
+    ///
+    /// For demo and typescript code, see:
+    /// - https://github.com/junomonster/nested-eip-712
+    /// - https://github.com/frangio/eip712-wrapper-for-eip1271
+    ///
+    /// Their nomenclature may differ from ours, although the high-level idea is similar.
+    ///
+    /// Of course, if you have control over the codebase of the wallet client(s) too,
+    /// you can choose a more minimalistic signature scheme like
+    /// `keccak256(abi.encode(address(this), hash))` instead of all these acrobatics.
+    /// All these are just for widespread out-of-the-box compatibility with other wallet clients.
+    function _erc1271HashForIsValidSignatureViaNestedEIP712(
```

Comment:
ok can you point me to the tests written on this in smart session validator please